Factor out numerical logic

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* #55692 Refactor C->C to C->R twice
* #55656 Clean up raise exception logic
* #55237 Enable fast gradcheck for real inputs and outputs
* #54480 Implement faster gradcheck but not enabled for most things
* **#54479 Factor out numerical logic**
* #54378 Refactor get numerical jacobian to calculate wrt all outputs at once

This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

Differential Revision: [D27728727](https://our.internmc.facebook.com/intern/diff/D27728727)

soulitzer committed:
Factor out numerical logic

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -252,8 +275,8 @@ def check_jacobians_equal(j1, j2, atol):
 
-def combine_jacobian_rows(jacobians_rows, inputs, output):
-    out_jacobians = allocate_jacobians_with_inputs(inputs, output.numel())
+def combine_jacobian_rows(jacobians_rows, inputs, dim):
```
nit: dim -> numel or "output_numel" might make more sense? `dim` is an overloaded word, it can mean "the number of dimensions in your tensor, e.g. tensor.dim()" or "the size of dimension X in your tensor, e.g. tensor.shape[X]"

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+                    jacobians_cols.append(None)
+
+
+def combine_jacobian_cols(jacobians_cols, outputs, input, dim):
```
nit: dim -> numel

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+                    jacobians_cols.append(None)
+
+
+def combine_jacobian_cols(jacobians_cols, outputs, input, dim):
```
Type hints would really help here (and in some of the other functions). E.g. it looks like `jacobian_cols` is always `Dict[int, List[Optional[Tensor]]` and this function returns a... `List[List[Tensor]]` ?

soulitzer replied:
In the process of adding type hints I realized that `List[Optional[Tensor]]]` should be `List[Tensor]` in many places.

soulitzer replied:
In the previous PR, I added a test for the specific code path that tests this (where input is real, but we return both complex and real outputs).

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+def compute_numerical_gradient(fn, entry, v, norm_v, nbhd_checks_fn):
+    # Performs finite differencing by perturbing `entry` in-place by `v` and
+    # returns the gradient of each of the outputs wrt to x at idx.
+    # we currently assume that the norm of delta equals eps
+    # assert(v == norm_v or v == (norm_v * 1j))
```
Where are `delta` and `eps` coming from? Should the assert be uncommented?

soulitzer replied:
Since we are computing norm_v inside the function now, lets just remove this assert entirely

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+    # we currently assume that the norm of delta equals eps
+    # assert(v == norm_v or v == (norm_v * 1j))
+
+    v = v.reshape(entry.shape) if isinstance(v, torch.Tensor) else v
```
`v` looks like it is always a Python number before this PR (I could be wrong), why did that change?

soulitzer replied:
Good catch, I shouldn't be doing this prematurely. In this PR v is still a python number. I'm moving the logic to handle tensors in the next PR.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+    return jacobians
+
+
+def compute_numerical_gradient(fn, entry, v, norm_v, nbhd_checks_fn):
```
Would it be simpler if we computed `norm_v` inside this function instead of passing it in as an argument? Depending on what `v` can be this is not too computationally expensive

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -180,68 +243,28 @@ def check_outputs_same_dtype_and_shape_in_neighborhood(output1, output2, idx, de
+
+        def jvp_fn(delta):
+            return compute_numerical_gradient(wrapped_fn, input_to_perturb, delta, eps, nbhd_checks_fn)
+        jacobian_cols[d_idx] = []
+        compute_numerical_jacobian_cols(jacobian_cols[d_idx], delta, jvp_fn, x.is_complex(), grad_out)
```
Mutation is difficult to reason about when there are a lot of variables flying around. Could we have something like the following?
```
jacobain_cols[d_idx] = compute_numerical_jacobian_cols(delta, jvp_fn, x.is_complex(), grad_out)
```

soulitzer replied:
Looks a lot cleaner with this change!

zou3519 reviewed (commented):
some initial comments/questions

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -141,11 +141,79 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+                if isinstance(grad_out, complex):
+                    # placeholder if grad_out is complex but output is not
+                    jacobians_cols.append(torch.zeros_like(ds_dx))
+                    continue
```
Previously we appended None here, but then I realized that line 212 wouldn't work properly (assigning NoneType to the column of a Tensor). The tests were still passing though because the code path was never reached, i.e., we didn't have any functions that take real input and have both complex and real output.

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -141,11 +141,79 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
+    if input_is_complex:            # C -> C, C -> R
+        ds_dy_tup = jvp_fn(delta * 1j)
+        for ds_dx, ds_dy in zip(ds_dx_tup, ds_dy_tup):
+            if not ds_dx.is_complex() and isinstance(grad_out, complex):
```
Added this here as well, because C-> R case was missed previously.

zou3519 replied:
Good catch!

zou3519 reviewed (approved):
lgtm, one suggestion

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]

soulitzer committed:
Update on "Factor out numerical logic"


This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

[ghstack-poisoned]
